src/main/java/forestry/core/data/ForestryRecipeProvider.java

```diff
@@ -0,0 +1,1050 @@
+				ForestryModuleUids.FACTORY);
+	}
+
+	private void registerFarmingRecipes(RecipeDataHelper helper) {
+		for (EnumFarmMaterial material : EnumFarmMaterial.values()) {
+			Item base = material.getBase().getItem();
+			helper.moduleConditionRecipe(
+					ShapedRecipeBuilder.shapedRecipe(FarmingBlocks.FARM.get(EnumFarmBlockType.PLAIN, material).block())
+							.key('#', base)
+							.key('C', CoreItems.ELECTRON_TUBES.get(EnumElectronTube.TIN).item())
+							.key('W', ItemTags.WOODEN_SLABS)
+							.key('I', ForestryTags.Items.INGOT_COPPER)
+							.patternLine("I#I").patternLine("WCW")
+							.addCriterion("has_copper", this.hasItem(ForestryTags.Items.INGOT_COPPER))::build,
+					ForestryModuleUids.FARMING);
+			helper.moduleConditionRecipe(
+					ShapedRecipeBuilder.shapedRecipe(FarmingBlocks.FARM.get(EnumFarmBlockType.GEARBOX, material).block())
+							.key('#', base)
+							.key('T', ForestryTags.Items.GEAR_TIN)
+							.patternLine(" # ").patternLine("TTT")
+							.addCriterion("has_tin_gear", this.hasItem(ForestryTags.Items.GEAR_TIN))::build,
+					ForestryModuleUids.FARMING);
+			helper.moduleConditionRecipe(
+					ShapedRecipeBuilder.shapedRecipe(FarmingBlocks.FARM.get(EnumFarmBlockType.HATCH, material).block())
+							.key('#', base)
+							.key('T', ForestryTags.Items.GEAR_TIN)
+							.key('D', ItemTags.WOODEN_TRAPDOORS)
+							.patternLine(" # ").patternLine("TDT")
+							.addCriterion("has_tin_gear", this.hasItem(ForestryTags.Items.GEAR_TIN))::build,
+					ForestryModuleUids.FARMING);
+			helper.moduleConditionRecipe(
+					ShapedRecipeBuilder.shapedRecipe(FarmingBlocks.FARM.get(EnumFarmBlockType.VALVE, material).block())
+							.key('#', base)
+							.key('T', ForestryTags.Items.GEAR_TIN)
+							.key('X', Tags.Items.GLASS)
+							.patternLine(" # ").patternLine("XTX")
+							.addCriterion("has_tin_gear", this.hasItem(ForestryTags.Items.GEAR_TIN))::build,
+					ForestryModuleUids.FARMING);
+			helper.moduleConditionRecipe(
+					ShapedRecipeBuilder.shapedRecipe(FarmingBlocks.FARM.get(EnumFarmBlockType.CONTROL, material).block())
+							.key('#', base)
+							.key('T', CoreItems.ELECTRON_TUBES.get(EnumElectronTube.GOLD).item())
+							.key('X', Tags.Items.DUSTS_REDSTONE)
+							.patternLine(" # ").patternLine("XTX")
+							.addCriterion("has_tin_gear", this.hasItem(ForestryTags.Items.GEAR_TIN))::build,
+					ForestryModuleUids.FARMING);
+		}
+	}
+
+	//TODO maybe I'm missing something, but this seems like the only reasonable way to do it...
+	private IngredientNBT createNbtIngredient(ItemStack stack) {
+		Constructor<IngredientNBT> constructor;
+		try {
+			constructor = IngredientNBT.class.getDeclaredConstructor(ItemStack.class);
+			constructor.setAccessible(true);
+			return constructor.newInstance(stack);
+
+		} catch (Exception e) {
+			throw new RuntimeException(e);
+		}
+
+	}
+
+
+	private void registerFluidsRecipes(RecipeDataHelper helper) {
```

Comment:
The constructor of `IngredientNBT` isn't public so I have to use reflection to get to it. Again this seems like a forge issue.